### PR TITLE
`multi_cn` recipes: `aishell2`, `magicdata`, `primewords`, `stcmds`, `tal_asr`, `tal_csasr`, `thchs_30`

### PR DESCRIPTION
**Pull request by shanguanma**
I want to add the multi_cn  recipe into icefall, so I first add the related dataset recipe.

### CODE REVIEW

File: lhotse/recipes/aishell2.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
+"""
+AISHELL2 (~1000 hours) if available(https://www.aishelltech.com/aishell_2).
```
Comment (csukuangfj) [inline lhotse/recipes/aishell2.py:2]: I recommend copying the English description of the dataset from
https://www.aishelltech.com/aishell_2
to here.

File: lhotse/recipes/aishell2.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    corpus_dir = Path(corpus_dir)
+    assert corpus_dir.is_dir(), f"No such directory: {corpus_dir}"
+    if output_dir is not None:
+        output_dir = Path(output_dir)
+        output_dir.mkdir(parents=True, exist_ok=True)
+
+    transcript_path = corpus_dir / "AISHELL-2" / "iOS" / "data" / "trans.txt"
+    transcript_dict = {}
+    with open(transcript_path, "r", encoding="utf-8") as f:
+        for line in f.readlines():
```
Comment (csukuangfj) [inline lhotse/recipes/aishell2.py:39]:
```suggestion
        for line in f:
```

File: lhotse/recipes/aishell2.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+            idx_transcript = line.split()
+            transcript_dict[idx_transcript[0]] = " ".join(idx_transcript[1:])
+
+    manifests = defaultdict(dict)
+    dataset_parts = ["train"]
+    for part in tqdm(
+        dataset_parts, desc="process aishell2 audio, it needs waste some time."
```
Comment (csukuangfj) [inline lhotse/recipes/aishell2.py:46]: Could you update the time information?

File: lhotse/recipes/aishell2.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    ):
+        logging.info(f"Processing aishell2 {part}")
+        # Generate a mapping: utt_id -> (audio_path, audio_info, speaker, text)
+        recordings = []
+        supervisions = []
+        wav_path = corpus_dir / "AISHELL-2" / "iOS" / "data" / "wav"
+        for audio_path in wav_path.rglob("**/*.wav"):
+            logging.info(f"Processing audio path  {audio_path}")
```
Comment (csukuangfj) [inline lhotse/recipes/aishell2.py:54]: Will it print lots of logs?

File: lhotse/recipes/aishell2.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+            idx = audio_path.stem
+            speaker = audio_path.parts[-2]
+            if idx not in transcript_dict:
+                logging.warning(f"No transcript: {idx}")
```
Comment (csukuangfj) [inline lhotse/recipes/aishell2.py:58]: I would suggest also printing out the `audio_path`.

File: lhotse/recipes/aishell2.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+    transcript_path = corpus_dir / "AISHELL-2" / "iOS" / "data" / "trans.txt"
+    transcript_dict = {}
+    with open(transcript_path, "r", encoding="utf-8") as f:
+        for line in f.readlines():
+            idx_transcript = line.split()
+            transcript_dict[idx_transcript[0]] = " ".join(idx_transcript[1:])
+
+    manifests = defaultdict(dict)
+    dataset_parts = ["train"]
```
Comment (csukuangfj) [inline lhotse/recipes/aishell2.py:44]: Are there `dev` and `test` datasets?

File: lhotse/recipes/aishell2.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
+"""
+AISHELL2 (~1000 hours) if available(https://www.aishelltech.com/aishell_2).
+"""
+
+import logging
+import os
+import shutil
+import tarfile
```
Comment (csukuangfj) [inline lhotse/recipes/aishell2.py:8]: Please remove unused imports.

**Review by csukuangfj — CHANGES_REQUESTED** [review]

**Comment by pzelasko** [comment]
Thanks, amazing work and lot's of new additions! Let me know when it is good to merge.

File: lhotse/recipes/tal_csasr.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+            idx = audio_path.stem
+            speaker = idx
+            if idx not in transcript_dict:
+                logging.warning(f"No transcript: {idx}")
+                continue
+            text = transcript_dict[idx]
+            if not audio_path.is_file():
+                logging.warning(f"No such file: {audio_path}")
+                continue
+            recording = Recording.from_file(audio_path)
```
Comment (pzelasko) [inline lhotse/recipes/tal_csasr.py:65]: You can speed this up by factor of `num_jobs` using `Recording.from_dir(wav_path, "*.wav", num_jobs=num_jobs)` to parallelize. You will still have to loop over transcript files.

File: lhotse/recipes/tal_csasr.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    transcript_dict = {}
+    with open(transcript_path, "r", encoding="utf-8") as f:
+        for line in f.readlines():
+            idx_transcript = line.split()
+            transcript_dict[idx_transcript[0]] = " ".join(idx_transcript[1:])
+
+    manifests = defaultdict(dict)
+    dataset_parts = ["train"]
+    for part in tqdm(
+        dataset_parts, desc="Process tal_csasr audio, it needs waste some time."
```
Comment (pzelasko) [inline lhotse/recipes/tal_csasr.py:47]: The "waste some time" description doesn't sound too informative.

File: lhotse/recipes/aishell2.py (hunk 1 above)
Comment (shanguanma) [reply lhotse/recipes/aishell2.py:2]: OK, I will change.

File: lhotse/recipes/aishell2.py (hunk 2 above)
Comment (shanguanma) [reply lhotse/recipes/aishell2.py:39]: OK

File: lhotse/recipes/aishell2.py (hunk 6 above)
Comment (shanguanma) [reply lhotse/recipes/aishell2.py:44]: Sorry, I have only a train set of aishell2.

File: lhotse/recipes/aishell2.py (hunk 5 above)
Comment (shanguanma) [reply lhotse/recipes/aishell2.py:58]: OK, I will change.

File: lhotse/recipes/aishell2.py (hunk 4 above)
Comment (shanguanma) [reply lhotse/recipes/aishell2.py:54]: It just shows that the program is running normally, it has taken the worry away.

File: lhotse/bin/modes/recipes/aishell2.py [hunk 10]
```diff
@@ -0,0 +1,15 @@
+import click
+
+from lhotse.bin.modes import download, prepare
```
[the remark this answers is not in the document]
Comment (shanguanma) [reply lhotse/bin/modes/recipes/aishell2.py:3]: Ok

File: lhotse/recipes/magicdata.py [hunk 11]
```diff
@@ -0,0 +1,146 @@
+"""
+Magicdata is an open-source Chinese Mandarin speech corpus by Magic Data Technology Co., Ltd.,
+containing 755 hours of scripted read speech data from 1080 native speakers of the Mandarin Chinese spoken
+in mainland China. The sentence transcription accuracy is higher than 98%.
+Publicly available on https://www.openslr.org/resources/68
```
[the remark this answers is not in the document]
Comment (shanguanma) [reply lhotse/recipes/magicdata.py:5]: OK, I see.

File: lhotse/recipes/magicdata.py [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+        recordings = []
+        supervisions = []
+        wav_path = corpus_dir / f"{part}"
+        for audio_path in wav_path.rglob("**/*.wav"):
+            idx = audio_path.stem
+            logging.info(f"process audio_path is {audio_path}")
+            speaker = audio_path.parts[-2]
+
+            if idx not in transcript_dict:
+                logging.warning(f"No transcript: {idx}")
```
[the remark this answers is not in the document]
Comment (shanguanma) [reply lhotse/recipes/magicdata.py:113]: OK

File: lhotse/recipes/tal_csasr.py (hunk 8 above)
Comment (shanguanma) [reply lhotse/recipes/tal_csasr.py:65]: OK, I will change.

File: lhotse/recipes/magicdata.py [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+from lhotse import validate_recordings_and_supervisions
+from lhotse.audio import Recording, RecordingSet
+from lhotse.supervision import SupervisionSegment, SupervisionSet
+from lhotse.utils import Pathlike, urlretrieve_progress
+
+
+def download_magicdata(
+    target_dir: Pathlike = ".",
+    force_download: Optional[bool] = False,
+    base_url: Optional[str] = "http://www.openslr.org/resources",
```
[the remark this answers is not in the document]
Comment (shanguanma) [reply lhotse/recipes/magicdata.py:29]: I will change.

File: lhotse/recipes/tal_csasr.py (hunk 8 above)
Comment (shanguanma) [reply lhotse/recipes/tal_csasr.py:65]: Recording don't from_dir method.

Comment (pzelasko) [reply lhotse/recipes/tal_csasr.py:65]: Right, I meant RecordingSet.from_dir

File: lhotse/recipes/aidatatang_200zh.py [hunk 14]
```diff
@@ -95,7 +97,10 @@ def prepare_aidatatang_200zh(
     manifests = defaultdict(dict)
     dataset_parts = ["dev", "test", "train"]
 
-    for part in dataset_parts:
+    for part in tqdm(
+        dataset_parts,
+        desc="process aishell audio, it needs waste about 2559 seconds time.",
```
Comment (csukuangfj) [inline lhotse/recipes/aidatatang_200zh.py:102]:
```suggestion
        desc="process aidatatang audio, it takes about 2559 seconds.",
```

File: lhotse/recipes/aishell2.py [hunk 15]
```diff
@@ -40,7 +40,7 @@ def prepare_aishell2(
     manifests = defaultdict(dict)
     dataset_parts = ["train"]
     for part in tqdm(
-        dataset_parts, desc="process aishell2 audio, it needs waste about 6 hours time."
+        dataset_parts, desc="Process aishell2 audio, it takes about 6 hours."
```
Comment (csukuangfj) [inline lhotse/recipes/aishell2.py:43]: I think 6 hours are too long for a 1000 hours dataset.

File: lhotse/recipes/tal_csasr.py (hunk 8 above)
Comment (csukuangfj) [reply lhotse/recipes/tal_csasr.py:66]: You can see its usage at
https://github.com/lhotse-speech/lhotse/blob/21b5b3f494893acc2b0ded291236194c803b851a/lhotse/recipes/mgb2.py#L148-L150

https://github.com/lhotse-speech/lhotse/blob/21b5b3f494893acc2b0ded291236194c803b851a/lhotse/recipes/mls.py#L101-L106

File: lhotse/recipes/aishell2.py (hunk 15 above)
Comment (shanguanma) [reply lhotse/recipes/aishell2.py:43]: Yes, I agree with you. I will implement it.

**Comment by pzelasko** [comment]
Is it good to merge @shanguanma @csukuangfj?

**Comment by csukuangfj** [comment]
+2

**Comment by pzelasko** [comment]
Thanks!

**Pull request merged**